doc/man/man4/sl-mux.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/sl-mux.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH SL-MUX 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.SH NAME
.B sl-mux
\- Signalling Link Multiplexer
.\"
.\"
.SH SYNOPSIS
.B #include <sys/stropts.h>
.br
.B #include <ss7/sli.h>
.br
.B #include <ss7/sl_ioctl.h>
.br
.B #include <ss7/sl_mux.h>
.HP
.BI "int " sl_stream " = open(" \fR/dev/sl\fP ", " \fRO_RDWR\fP );
.\"
.\"
.SH DESCRIPTION
.PP
.B sl-mux
is a Signalling Link multiplexing driver.  The purpose of the driver is to consolidate signalling
link streams from a wide range of drivers and provide common naming, addressing, management and
monitoring procedures for signalling links.
.PP
.B sl-mux
is used in conjunction with a Signalling Link configuration utility and the Signalling Link
configuration daemon,
.BR slconfd (8).
.PP
The configuration utility is responsible for opening, configuring, and attaching Signalling Link
streams from an associated drivers and linking them under the
.B sl-mux
multiplexing driver.  This makes signalling links from a number of drivers available for use through
the single
.B sl-mux
multiplexing driver.  Configuration utilities exist for each type of Signalling Link provider.  For
example, the
.BR x400p-ss7 (4)
driver uses a configuration utility called
.BR x400pconfig (8);
.BR m2pa (4),
.BR m2paconf (8);
etc.
.PP
The Signalling Link configuration daemon,
.BR slconfd (8),
is responsible for the general configuration of Signalling Links and performing common
administration (such as management, configuration, monitoring, statistics collection and event
reporting).
.PP
The facilities provided by the
.B sl-mux
multiplexing driver are as follows:
.\"
.SS Addressing
.PP
The
.B sl-mux
provides a global addressing mechanism for addressing signalling links.  There
are three addressing mechanisms provided by the driver:
.TP
.B Global PPA
A Physical Point of Attachment addressing mechanism that is global to the
host.  Each signalling link stream is assigned a 32-bit PPA that is unique to
the host platform.  There is no restrictions on the internal form of the PPA,
merely that it be unique for each signalling link.  The
.BR slconf (5)
configuration file provides a mechanism whereby the system administrator can
assign arbitrary PPA addresses to signalling links from multiple drivers.
.TP
.B CLEI
A Common Language Equipment Identifier (ascii string) addressing mechanism that
is global across all platforms.  This is the Telcordia/Bellcore standard
mechanism for naming equipment.  The driver does not impose a restriction on
the CLEI string, only that the string be unique for each signalling link, and
that the string be no longer than 32 characters in length.
.TP
.B Minor Device Number
The minor device numbers 11 through 4106 can be opened an obtain individual
Style I SL Streams.  These minor device number correspond to device 1 through
4096 (simply add 10 to the device number to arrive at the character minor
device number).
.PP
The
.B sl-mux
driver permits Style I signalling link streams to be opened; or, Style II
signalling link streams to be opened and attached to specific signalling links
using either the PPA or CLEI as an address for attachment.  See
.BR LMI_ATTACH_REQ (7).
Style I Streams are obtained by opening character minor device number 11
through 4106 corresponding to unit 1 through 4096.
Style II Streams are obtained by opening character minor device number zero
.RB ( /dev/sl ),
which is a clone minor device.
.PP
Linking of signalling link streams beneath the multiplexing driver is
performed by the Signalling Link configuration daemon,
.BR slconfd (8).
The daemon opens the administrative minor device,
.IR /dev/sl-admin ,
and then opens signalling link streams on various drivers according to the
configuration specified in a configuration file,
.BR slconf (5),
attaches and configures the streams.  The daemon then links the SL streams
under the
.B sl-mux
driver using the
.IR I_PLINK (7)
input-output control.  It then uses the
.B SLM_IOCSPPA
input-output control to associate a global PPA and CLEI with the linked
stream.  The daemon does this for each and every signalling link stream that
is to be linked under the driver.
.\"
.SS Configuration
.PP
The
.B sl-mux
driver provides the ability to perform configuration and administration of
signalling links.  This is accomplished by opening a specialized minor device
number of the driver,
.IR /dev/sl-admin .
Using modified
.BR sl_ioctl (4)
signalling link input-output controls, the administrative application can
configure and query the configuration for each and any (or all) signalling
links.  This interface is used by the the
.BR slconfd (8)
configuration utility to query and configure signalling links in accordance
with the signalling link management model presented in Q.751-1.
.[
q751-1
.]
.PP
Two forms of configuration are possible: public and private configuration:
.IP 1. \w'0.\(em'u
In the public form, streams are opened by the configuration utility, normally
as part of a system startup procedure, and linked under the
.B sl-mux
driver using the
.IR I_PLINK (7)
input-output control operation.  Signalling link streams linked in this
fashion are available for attachment by any opened upper stream.
Also, these signalling links exhibit persistence as described below.
.IP 2.
In the private form, an upper stream is opened by an application and a
signalling link is configured and linked under the stream using the
.IR I_LINK (7)
input-output control operation.  The upper stream is attached to this lower
stream and the signalling link is not available for attachment by other upper
streams.  When the control stream is closed, the lower signalling link stream
is also closed as part of the dismantling of the multiplex.  Streams linked in
this fashion are private to the configuring application for use, but are
available for the collection of statistics, reporting of events, and
signalling link monitoring.
.\"
.SS Persistence
.PP
The
.B sl-mux
provides the ability for a Signalling Link to exist and operate without an
application present.  When configured, Signalling Links may be persistently
linked beneath the
.B sl-mux
driver using the
.IR I_PLINK (7)
input-output control operation.  Signalling Link linked in this fashion can be
attached by Streams opened on the upper multiplex and brought into service.
The
.B sl-mux
driver tracks the state of the Signalling Link and, if the upper Stream is
closed while the link is operating, the
.B sl-mux
driver places the link into local processor outage.  When a stream is again
opened, attached and enabled, local processor recovery is indicated and the
Signalling Link and be resumed without realignment.  This is a useful feature
for reconfiguring (or replacing) MTP on a live system, for example, to perform
live upgrades.
.\"
.SS Statistics
.PP
The
.B sl-mux
driver provides the ability to collect statistics (operational measurements)
for each and any signalling link.  This is accomplished by opening a
specialized minor device number on the driver,
.IR /dev/sl-stats .
Using modified
.BR sl_ioctl (4)
signalling link input-output controls, the administrative application can
collect statistics for each and any (or all) signalling links.  This interface
is used by the
.BR slstats (8)
utility to collect and record operational measurements for signalling links in
accordance with Q.751-1.
.[
q.751-1
.]
.\"
.SS Events
.PP
The
.B sl-mux
driver provides the ability to collect event notifications (operational
measurements) for each and any signalling link.  This is accomplished by
opening a specialized minor device number on the driver,
.IR /dev/sl-events .
Using modified
.BR sl_ioctl (4)
signalling link input-output controls, the administrative application can
collect event notifications for each and any (or all) signalling links.  This
interface is used by the
.BR slstats (8)
utility to collect and record operational measurements for signalling links in
accordance with Q.751-1.
.[
q751-1
.]
.\"
.SS Monitoring
.PP
The
.B sl-mux
driver provides the ability to monitor the service interface to each or any
signalling link.  This is accomplished by opening a specialized minor device
number on the driver,
.IR /dev/sl-mon ,
and attaching it using the
.BR SLM_IOCSMON " and " SLM_IOCCMON
input-output controls to specify the Signalling Links by lower multiplex
index, gloabl PPA or CLEI that the monitoring Stream wishes to monitor.
The stream may also monitor all lower stream by using a zero lower multipelx
index, zero global PPA and null CLEI.  Once activated for monitoring, lower
streams will deliver a copy of each and every STREAMS message that is passed
between an upper SL-user and the monitored signalling link(s), regardless
of whether the primiitve is an SL-user to SL-provider primitive or an
SL-provider to SL-user primitive.  The STREAMS message will be encapsulated
as follows:
.PP
The control part of the message will be copied and encapsulated in a band zero
(0)
.BR M_PROTO (9)
message block with a
.B slmux_mon
structure prepended as follows:
.sp
.nf
\fC\s-1\
struct slmux_mon {
    struct slmux_ppa mon_ppa;
    int mon_dir;
    uchar mon_msg_type;
    uchar mon_msg_band;
    ushort mon_msg_flags;
    uchar mon_msg_data[0];
};
\s+1\fP
.fi
.PP
The
.B slmux_mon
structure is defined in the
.RB < ss7/sl_mux.h >
header file and contains the following fields:
.TP
.I mon_ppa

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/sl-mux.c
/*****************************************************************************

 @(#) File: src/drivers/sl-mux.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/sl-mux.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  This is a signalling link multiplexer: its primary purpose is to provide a single nexus for the
 *  convenient arrangement and handling of signalling link interfaces.  Signalling links are opened
 *  on other drivers and attached to a specific PPA using the procedures associated with that
 *  device.  The signalling link stream is then linked under this multiplexing driver using the
 *  I_PLINK input-output control.  Once linked, configuration input-output controls specific to this
 *  driver are used to assign a global PPA and a CLEI to the signalling link.  Streams opened on
 *  this SL-MUX driver can then attach to any defined global PPA or CLEI which results in
 *  atttachment of the upper multiplex stream to the appropriate lower multiplex stream.
 */

/**
 * sl_w_ioctl: - process M_IOCTL message
 * @q: active queue (upper write queue)
 * @mp: the message
 *
 * Linking of streams: streams are linked under the multiplexing driver by opening an upper stream
 * and then linking a signalling link stream under the multiplexing driver.  Then the SL_SETLINK
 * input-output control is used with the multiplexer id to set the global-PPA and CLEI associated
 * with the signalling link.  The SL_GETLINK input-output control can be used at a later date to
 * determine the multiplexer id for a given signalling link stream.
 */
static struct int
sl_w_ioctl(queue_t *q, mblk_t *mp)
{
	struct iocblk *ioc = (struct iocblk *) mp->b_rptr;

	switch (ioc->ioc_cmd) {
	case I_LINK:
	case I_PLINK:
	{
		struct linkblk *l;

		if (!mp->b_cont)
			mi_copy_done(q, mp, EINVAL);

		l = (struct linkblk *) mp->b_cont->b_rptr;

		if (!(bot = kmem_alloc(sizeof(*bot), KM_NOSLEEP)))
			mi_copy_done(q, mp, ENOMEM);

		write_lock_str(&mux_lock, flags);
		bot->next = mux_links;
		bpt->prev = &mux_links;
		mux_links = bot;
		bot->dev = l->l_index;
		bot->rq = RD(l->l_qtop);
		bot->wq = l->l_qtop;
		bot->other = NULL;
		noenable(bot->rq);
		l->l_qtop->q_ptr = RD(l->l_qtop)->q_ptr = (void *) bot;
		write_unlock_str(&mux_lock, flags);
		mi_copy_done(q, mp, 0);
		return (0);
	}
	case I_UNLINK:
	case I_PUNLINK:
	{
		struct linkblk *l;

		if (!mp->b_cont)
			mi_copy_done(q, mp, EINVAL);

		l = (struct linkblk *) mp->b_cont->b_rptr;

		write_lock_str(&mux_lock, flags);
		for (bot = mux_list; bot; bot = bot->next)
			if (bot->dev == l->l_index)
				break;
		if (!bot) {
			write_unlock_str(&mux_lock, flags);
			mi_copy_done(q, mp, EINVAL);
			return (0);
		}
		/* Note that the lower multiplex driver put and service procedures must be prepared
		   to be invoked event after the M_IOCACK for the I_UNLINK or I_PUNLINK ioctl has
		   been returned.  THis is because the setq(9) back to the Stream head is not
		   performed until after the acknowledgement has been received.  We set q->q_ptr to
		   a null multiplex structure to keep the lower Stream functioning until the setq(9)
		   is performed. */
		l->l_qtop->q_ptr = RD(l->l_qtop)->q_ptr = &no_mux;
		if ((*bot->prev = bot->next)) {
			bot->next = NULL;
			bot->prev = &bot->next;
		}
		bot->other = NULL;
		kmem_free(bot, sizeof(*bot));
		/* hang up all upper streams that feed this lower stream */
		for (top = mux_opens; top; top = top->next) {
			if (top->other == bot) {
				putnextctl(top->rq, M_HANGUP);
				top->other = NULL;
			}
		}
		write_unlock_str(&mux_lock, flags);
		mi_copy_done(q, mp, 0);
		return (0);
	}
	case SL_SETLINK:
	{
		struct sl_mux_ppa *sm;

		/* This input-output control is used to set the global-PPA and CLEI associated with
		   a lower multiplex stream.  The argument is an sl_mux_ppa structure that contains
		   the multiplex id, the 32-bit PPA, and a CLEI string of up to 32 characters in
		   length. */
		mi_copyin(q, mp, NULL, sizeof(struct sl_mux_ppa));
		return (0);
	}
	case SL_GETLINK:
	{
		/* This input-output control is used to obtain the multiplex-id assocated with a
		   lower multiplex stream.  The argument is an sl_mux_ppa structure that contains a
		   32-bit PPA or CLEI string of up to 32 characters in length.  It returns the
		   multiplex id in the same structure. */
		mi_copyin(q, mp, NULL, sizeof(struct sl_mux_ppa));
		return (0);
	}
	default:
		if (mux->other && mux->other->rq) {
			if (bcanputnext(mux->other->rq, mp->b_band)) {
				putnext(mux->other->rq, mp);
				return (0);
			}
			return (-EBUSY);
		}
		break;
	}
	mi_copy_done(q, mp, EINVAL);
	return (0);
}

static struct int
sl_w_iocdata(queue_t *q, mblk_t *mp)
{
	struct copyresp *cp = (struct copyresp *) mp->b_rptr;
	struct sl_mux_ppa sm;
	unsigned long flags;
	struct sl_mux *bot, *mux;
	mblk_t *dp, *bp;

	switch (cp->cp_cmd) {
	case SL_SETLINK:
		switch (mi_copy_state(q, mp, &dp)) {
		case MI_COPY_CASE(MI_COPY_IN, 1):
			/* This input-output control is used to set the global-PPA and CLEI
			   associated with a lower multiplex stream.  The argument is an sl_mux_ppa 
			   structure that contains the multiplex id, the 32-bit PPA, and a CLEI
			   string of up to 32 characters in length. */
			bcopy(dp->b_rptr, &sm, sizeof(sm));
			/* The assigned global PPA must be non-zero. */
			if (sm.sm_ppa == 0) {
				mi_copy_done(q, mp, EINVAL);
				return (0);
			}
			/* The assigned CLEI must be non-null. */
			if (sm.sm_clei[0] == '\0') {
				mi_copy_done(q, mp, EINVAL);
				return (0);
			}
			write_lock_str(&mux_lock, flags);
			/* Find the stream with the associated MUXID. */
			for (bot = mux_list; bot; bot = bot->next)
				if (bot->dev == sm.sm_index)
					break;
			if (!bot) {
				write_unlock_str(&mux_lock, flags);
				mi_copy_done(q, mp, EINVAL);
				return (0);
			}
			/* The global PPA to assign must be unique. */
			for (mux = mux_list; mux; mux = mux->next)
				if (mux != bot && mux->ppa == sm.sm_ppa)
					break;
			if (mux) {
				write_unlock_str(&mux_lock, flags);
				mi_copy_done(q, mp, EINVAL);
				return (0);
			}
			/* The CLEI to assign must also be unique. */
			for (mux = mux_list; mux; mux = mux->next)
				if (mux != bot && strncmp(mux->clei, sm.sm_clei, 32) == 0)
					break;
			if (mux) {
				write_unlock_str(&mux_lock, flags);
				mi_copy_done(q, mp, EINVAL);
				return (0);
			}
			bot->ppa = sm.sm_ppa;
			bcopy(bot->clei, sm.sm_clei, 32);
			write_unlock_str(&mux_lock, flags);
			mi_copy_done(q, mp, 0);
			return (0);
		case -1:
			return (0);
		}
		break;
	case SL_GETLINK:
		switch (mi_copy_state(q, mp, &dp)) {
		case MI_COPY_CASE(MI_COPY_IN, 1):
			/* This input-output control is used to obtain the multiplex-id assocated
			   with a lower multiplex stream.  The argument is an sl_mux_ppa structure
			   that contains a 32-bit PPA or CLEI string of up to 32 characters in
			   length.  It returns the multiplex id in the same structure. */
			bcopy(dp->b_rptr, &sm, sizeof(sm));
			if (sm.sm_ppa != 0) {
				/* If the supplied global PPA is non-zero, then search on PPA. */
				read_lock_str(&mux_lock, flags);
				for (bot = mux_list; bot; bot = bot->next)
					if (bot->ppa == sm.sm_ppa)
						break;
				if (!bot) {
					read_unlock_str(&mux_lock, flags);
					mi_copy_done(q, mp, EINVAL);
					return (0);
				}
				strncpy(sm.sm_clei, bot->clei, 32);
				sm.sm_index = bot->dev;
				read_unlock_str(&mux_lock, flags);
				if ((bp = mi_copyout_alloc(q, mp, NULL, sizeof(sm)))) {
					bcopy(&sm, bp->b_rptr, sizeof(sm));
					mi_copyout(q, mp);
				}
				return (0);
			} else if (sm.sm_clei[0] != '\0') {
				/* If the supplied CLEI is non-null, then search on CLEI. */
				read_lock_str(&mux_lock, flags);
				for (bot = mux_list; bot; bot = bot->next)
					if (strncmp(sm.sm_clei, bot->clei, 32) == 0)
						break;
				if (!bot) {
					read_unlock_str(&mux_lock, flags);
					mi_copy_done(q, mp, EINVAL);
					return (0);
				}
				sm.sm_ppa = bot->ppa;
				sm.sm_index = bot->dev;
				read_unlock_str(&mux_lock, flags);
				if ((bp = mi_copyout_alloc(q, mp, NULL, sizeof(sm)))) {
					bcopy(&sm, bp->b_rptr, sizeof(sm));
					mi_copyout(q, mp);
				}
				return (0);
			} else {
				/* If PPA is zero and CLEI is null, return the first item in the
				   list. */
				read_lock_str(&mux_lock, flags);
				bot = mux_list;
				if (!bot) {
					read_unlock_str(&mux_lock, flags);
					mi_copy_done(q, mp, EINVAL);
					return (0);
				}
				sm.sm_ppa = bot->ppa;
				strncpy(sm.sm_clei, bot->clei, 32);
				sm.sm_index = bot->dev;
				read_unlock_str(&mux_lock, flags);
				if ((bp = mi_copyout_alloc(q, mp, NULL, sizeof(sm)))) {
					bcopy(&sm, bp->b_rptr, sizeof(sm));
					mi_copyout(q, mp);
				}
				return (0);
			}
		case MI_COPY_CASE(MI_COPY_OUT, 1):
			mi_copy_done(q, mp, 0);
			return (0);
		case -1:
			return (0);
		}
		break;
	default:
		if (mux->other && mux->other->rq) {
			if (bcanputnext(mux->other->rq, mp->b_band)) {
				putnext(mux->other->rq, mp);
				return (0);
			}
			return (-EBUSY);
		}
		break;
	}
	mi_copy_done(q, mp, EINVAL);
	return (0);
}

static int
sl_w_msg(queue_t *q, mblk_t *mp)
{
	struct sl *mux = SL_PRIV(q);

	switch (DB_TYPE(mp)) {
	case M_IOCTL:
		return sl_w_ioctl(mux, q, mp);
	case M_IOCDATA:
		return sl_w_iocdata(mux, q, mp);
	}
	if (mux->other && mux->other->rq) {
		if (bcanputnext(mux->other->rq, mp->b_band)) {
			putnext(mux->other->rq, mp);
			return (0);
		}
		return (-EBUSY);
	}
	freemsg(mp);
	return (0);
}

static int
sl_r_msg(queue_t *q, mblk_t *mp)
{
	struct sl *mux = SL_PRIV(q);

	if (mux->other && mux->other->rq) {
		if (bcanputnext(mux->other->rq, mp->b_band)) {
			putnext(mux->other->rq, mp);
			return (0);
		}
		return (-EBUSY);
	}
	freemsg(mp);
	return (0);
}

static streamscall int
sl_uwput(queue_t *q, mblk_t *mp)
{
	if ((!pcmsg(DB_TYPE(mp)) && (q->q_first || (q->q_flag & QSVCBUSY))) || sl_w_msg(q, mp))
		putq(q, mp);
	return (0);
}
static streamscall int
sl_uwsrv(queue_t *q)
{
	mblk_t *mp;

	while ((mp = getq(q))) {
		if (sl_w_msg(q, mp)) {
			putbq(q, mp);
			break;
		}
	}
	return (0);
}
static streamscall int
sl_ursrv(queue_t *q)
{
	struct sl *mux = SL_PRIV(q);

	if (mux->other && mux->other->rq)
		qenable(mux->other->rq);
}
static streamscall int
sl_lwsrv(queue_t *q)
{
	struct sl *mux = SL_PRIV(q);

	if (mux->other && mux->other->wq)
		qenable(mux->other->wq);
}
static streamscall int
sl_lrsrv(queue_t *q)
{
	mblk_t *mp;

	while ((mp = getq(q))) {
		if (sl_r_msg(q, mp)) {
			putbq(q, mp);
			break;
		}
	}
	return (0);
}
static streamscall int
sl_lrput(queue_t *q, mblk_t *mp)
{
	if ((!pcmsg(DB_TYPE(mp)) && (q->q_first || (q->q_flag & QSVCBUSY))) || sl_r_msg(q, mp))
		putq(q, mp);
	return (0);
}

static struct qinit sl_urinit = {
	.qi_srvp = sl_ursrv,
	.qi_qopen = sl_qopen,
	.qi_qclose = sl_qclose,
	.qi_minfo = &sl_minfo,
	.qi_mstat = &sl_urstat,
};
static struct qinit sl_uwinit = {
	.qi_putp = sl_uwput,
	.qi_srvp = sl_uwsrv,
	.qi_minfo = &sl_minfo,
	.qi_mstat = &sl_uwstat,
};
static struct qinit sl_lrinit = {
	.qi_putp = sl_lrput,
	.qi_srvp = sl_lrsrv,
	.qi_minfo = &sl_minfo,
	.qi_mstat = &sl_lrstat,
};
static struct qinit sl_lwinit = {
	.qi_srvp = sl_lwsrv,
	.qi_minfo = &sl_minfo,
	.qi_mstat = &sl_lwstat,
};

static struct streamtab sl_info = {
	.st_rdinit = &sl_urinit,
	.st_wrinit = &sl_uwinit,
	.st_muxrinit = &sl_lrinit,
	.st_muxwinit = &sl_lwinit,
};

static struct cdevsw sl_cdev = {
	.d_name = CONFIG_STREAMS_SL_MUX_NAME,
	.d_str = &sl_info,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

static __init int
sl_mux_init(void)
{
	if ((err = register_strdev(&sl_cdev, major)) < 0)
		return (err);
	if (major == 0)
		major == err;
	return (0);
}

static __exit void
sl_mux_exit(void)
{
	unregister_strdev(&sl_cdev, major);
}

module_init(sl_mux_init);
module_exit(sl_mux_exit);
